### ISSUE TITLE
Link mapping with a dynamic subject fails

### ISSUE THREAD

**Issue by mabounassif** [morph-kgc/morph-kgc#254]
####  What Happens?
Link mapping with a dynamic subject fails with the following error: `KeyError: 'subject'`

#### To Reproduce
```yml
prefixes:
  kb-ingredientname: http://idea.rpi.edu/heals/kb/ingredientname/
  owl: http://www.w3.org/2002/07/owl#
  usda: http://idea.rpi.edu/heals/kb/usda#
  nutriment: http://idea.rpi.edu/heals/kb/nutriment/
  ex: http://example.com/
  grel: http://users.ugent.be/~bjdmeest/function/grel.ttl#
  idlab-fn: https://w3id.org/imec/idlab/function#
  morph-kgc: https://github.com/morph-kgc/morph-kgc/function/built-in.ttl#
  rdf: http://www.w3.org/1999/02/22-rdf-syntax-ns#

sources:
  nutr_def_source:
    access: knowledge/usda-nutrition/NUTR_DEF.csv
    referenceFormulation: csv
  nutr_data_source:
    access: knowledge/usda-nutrition/NUT_DATA.csv
    referenceFormulation: csv

mappings:
  nutriment_amount:
    sources:
      - nutr_data_source
    s:
      - function: morph-kgc:uuid
    po:
      - p: rdf:type
        o:
          mapping: nutriment
          condition:
            function: equal
            parameters:
              - [str1, $(Nutr_No), s]
              - [str2, $(Nutr_No), o]

  nutriment:
    sources:
      - nutr_def_source
    s: usda:$(Nutr_No)
    po:
      - p: rdf:type
        o: nutriment:$(NutrDesc)

```

#### Environment (please complete the following information):
 - OS: Ubuntu 22.04 LTS
 - Python version: 3.10
 - Morph-KGC version: 2.7.0

**Comment by arenas-guerrero-julian** [comment]
Hi @mabounassif ,

I need more information, could you please provide the entire stack trace?

Also, could you please try if the mapping without the `nutrient_amount` triples map works?

**Comment by mabounassif** [comment]
Absolutely, 

```bash
Exception has occurred: KeyError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
'subject'
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 375, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/__init__.py", line 49, in materialize_set
    triples = set().union(*pool.starmap(_materialize_mapping_group_to_set,
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/__init__.py", line 65, in materialize
    triples = materialize_set(config, python_source)
  File "/app/knowledge/knowledge.py", line 23, in generate_rdf_graph
    assertions = morph_kgc.materialize(config_path)
  File "/app/knowledge/knowledge.py", line 86, in <module>
    generate_rdf_graph(sha=sha)
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
KeyError: 'subject'
```

Something like the following works without any issues:

```yaml
prefixes:
  kb-ingredientname: http://idea.rpi.edu/heals/kb/ingredientname/
  owl: http://www.w3.org/2002/07/owl#
  usda: http://idea.rpi.edu/heals/kb/usda#
  nutriment: http://idea.rpi.edu/heals/kb/nutriment/
  ex: http://example.com/
  grel: http://users.ugent.be/~bjdmeest/function/grel.ttl#
  idlab-fn: https://w3id.org/imec/idlab/function#
  morph-kgc: https://github.com/morph-kgc/morph-kgc/function/built-in.ttl#
  rdf: http://www.w3.org/1999/02/22-rdf-syntax-ns#

sources:
  nutr_def_source:
    access: knowledge/usda-nutrition/NUTR_DEF.csv
    referenceFormulation: csv
  nutr_data_source:
    access: knowledge/usda-nutrition/NUT_DATA.csv
    referenceFormulation: csv

mappings:
  nutriment_amount:
    sources:
      - nutr_data_source
    s:
      - usda:$(NDB_No)
    po:
      - p: rdf:type
        o:
          mapping: nutriment
          condition:
            function: equal
            parameters:
              - [str1, $(Nutr_No), s]
              - [str2, $(Nutr_No), o]

  nutriment:
    sources:
      - nutr_def_source
    s: usda:$(Nutr_No)
    po:
      - p: rdf:type
        o: nutriment:$(NutrDesc)
```

Also removing `nutriment_amount` works as well.

**Comment by mabounassif** [comment]
Thanks for looking into this @arenas-guerrero-julian, I have attached samples of the data I'm using
[NUTR_DEF-10.csv](https://github.com/user-attachments/files/15523698/NUTR_DEF-10.csv)
[NUT_DATA-10.csv](https://github.com/user-attachments/files/15523699/NUT_DATA-10.csv)

**Comment by mabounassif** [comment]
Disabling multiprocessing leads to the following error stack:

```bash
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 3805, in get_loc
    return self._engine.get_loc(casted_key)
  File "index.pyx", line 167, in pandas._libs.index.IndexEngine.get_loc
  File "index.pyx", line 196, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 7081, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 7089, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'subject'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/root/.vscode-server/extensions/ms-python.debugpy-2024.6.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/app/knowledge/knowledge.py", line 86, in <module>
    generate_rdf_graph(sha=sha)
  File "/app/knowledge/knowledge.py", line 23, in generate_rdf_graph
    assertions = morph_kgc.materialize(config_path)
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/__init__.py", line 65, in materialize
    triples = materialize_set(config, python_source)
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/__init__.py", line 57, in materialize_set
    triples.update(_materialize_mapping_group_to_set(mapping_group, rml_df, fnml_df, config, python_source))
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/materializer.py", line 418, in _materialize_mapping_group_to_set
    data = _materialize_rml_rule(rml_rule, rml_df, fnml_df, config, python_source=python_source)
  File "/usr/local/lib/python3.10/site-packages/morph_kgc/materializer.py", line 395, in _materialize_rml_rule
    data['triple'] = data['subject'] + ' ' + data['predicate'] + ' ' + data['object']
  File "/usr/local/lib/python3.10/site-packages/pandas/core/frame.py", line 4102, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/usr/local/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 3812, in get_loc
    raise KeyError(key) from err
KeyError: 'subject'
```

**Comment by arenas-guerrero-julian** [comment]
Thanks for looking into this @mabounassif .

You are right, and we should have it not only for subjects, but also for predicates and objects.

**Comment by mabounassif** [comment]
@arenas-guerrero-julian , I created 2 PRs.

On master, the test case that captures this issue is passing (PR #256).

For version 2.7.0, it's not and requires code change. If you plan on backporting the fix to 2.7.0 let me know and I will change the target of the PR #255 to the branch that you'd create.

**Comment by arenas-guerrero-julian** [comment]
Thanks for the PR @mabounassif ,

My question is that the test case is passing without any changes to the code, am I right?

I will create v2.8.0 as soon as we are done with this issue 🙂

**Comment by mabounassif** [comment]
Hello @arenas-guerrero-julian,

the master branch was working without any code chanage. PR #256 was just adding a test case to master. 

I made another PR that branched out off the 2.7.0 tag PR #255 which wasn't supposed to be merged into master.

So I think right now there's redundancy in master and this is no longer needed: https://github.com/morph-kgc/morph-kgc/blob/main/src/morph_kgc/materializer.py#L302.

You can safely remove this line and luckily the tests all pass!

**Comment by arenas-guerrero-julian** [comment]
This makes sense because I refactored `materializer.py` after v2.7.0. I just published v2.8.0 🙂.
